### ISSUE TITLE
REGRESSION in FO4Edit 4.0.4: Keyword from base mod not resolved correctly in overridden Armor record

### ISSUE THREAD

**Issue by RealHandcuffs** [TES5Edit/TES5Edit#1078]
**What is the version of xEdit you are using?**
FO4Edit 4.0.4 (5B1C2C51)

**Describe the bug**
- I have a mod that declares an Armor record in an ESP
- there is an optional compatibility patch that overrides the record in an ESL-flagged ESP, adding keywords
- when opening the patch in FO4Edit 4.0.4, one of the keywords is messed up, it shows up as <Warning: Could not be resolved, but is possibly hardcoded in the engine>
- when opening the patch in FO4Edit 4.0.3, the keyword is resolved correctly

**To reproduce**
1. Install dependencies for patch:
    - [Torture Devices](https://www.loverslab.com/files/file/3341-torturedevices/)
    - [Devious Devices RC8](https://www.loverslab.com/topic/73925-devious-devices/page/137/#comment-3572600)
2. Install [RealHandcuffs](https://www.loverslab.com/files/file/8485-real-handcuffs/), choosing to install the optional patch for Devious Devices
3. Start FO4Edit and wait for the loader to finish
4. Navigate to the patch (RealHandcuffs_DD_Compatibility.esp or RealHandcuffs_DD_Servitron_Compatibility.esp)
5. Select one of the RH_ Armor items, e.g. RH_Handciffs

**Expected behavior**
FO4Edit shows the keyword RH_Restraint as unchanged in the patch.

**Actual behavior**
FO4Edit shows the keyword as removed, and a new keyword which if cannot resolve.

**Screenshots**
Bad behavior in FO4Edit 4.0.4:
![bad_4_0_4](https://user-images.githubusercontent.com/65441098/219128042-ef8f8cbf-060c-4de5-919d-b2f455760c1a.png)
Correct behavior in FO4Edit 4.0.3:
![good_4_0_3](https://user-images.githubusercontent.com/65441098/219128019-ac52f0fa-17cc-41a3-8505-8b27a00e8968.png)

**Additional context**
According to [my tests](https://www.loverslab.com/topic/118151-real-handcuffs/?do=findComment&comment=4001218), the game considers the keyword as unchanged, i.e. the situation in-game is as shown by FO4Edit 4.0.3.

**Comment by jbostrus** [comment]
So...This isn't a bug in 4.0.4. It's actually a bug in 4.0.3 in that it shows it resolving.

The game could actually misbehave when used in the way it is.

Change the HEDR version in the file header of the compatibility patch plugin to 1.0 instead of 0.95.

When the HEDR version is 0.95 it specifies to the game exe that any localized FormID that is less than 0x800 is a hard coded form. So the keyword you're working with `xx000009` is read as 00000009 not B3000009.

**Comment by RealHandcuffs** [comment]
@jbostrus I don't know anything about HEDR versions, but can you please clarify two things for me?

1. All other keywords from the base (like RH_ma_handcuffs) are resolved correctly, why?
2. Are you sure that it will not work in-game? Because I [tested ](https://www.loverslab.com/topic/118151-real-handcuffs/page/58/#comment-4001218) the behavior in-game before coming here, and my test points to the game resolving the keyword.

**Comment by jbostrus** [comment]
RE 1: Because they're all higher than 0x800+
RE 2: It **could** misbehave **if** the actual ID it happens to be would conflict with a real base record in the 0x001 - 0x7FF range. I don't think 0x9 is used, but if it happened to be B3000007 then it would probably make the game crash. Because it would be trying to override the player NPC record with the keyword. Not add a new keyword.

**Comment by VlitS** [comment]
For header version/file header -> Click the plugin name in xEdit, it will show the file header. Then look under the section that says 
**HEDR - Header**
To use an ID of xxx009 like it is using, RealHandcuffs.esp MUST have a file version of 1.000000.

If you have issues like this in the future, it is best to get help on the discord (discord link in the top right of xEdit), and if we confirm it is a bug in xEdit rather than a problem with a plugin, that is when you would want to submit an issue here.

**Comment by jbostrus** [comment]
So basically because it just happens to be 0x9 you got lucky nothing bad is happening (right now). The problem of leaving it like it is, is that if another mod also did that you now have what is essentially 2 mods trying to inject conflicting records.

So to properly fix the problem there's 2 options.

- The least problem inducing is to change the HEDR version of plugins that involve that record to 1.0. The drawback here is that the users of the plugin need to be on a current game exe (1.10.162+) which then excludes VR.
- The more problematic option is to renumber any records with an ID less than 0x800. Then all related plugins could remain HEDR version 0.95. I would argue against this only because you could in the future end up referencing to another plugin whose record(s) may have an ID less than 0x800. And then you're in the same boat again needing to change the HEDR version.

I made a [F4SE dll](https://www.nexusmods.com/fallout4/mods/42497) to allow the older game exes like VR to load a plugin with the HEDR 1.0, but the problem of it forcing the lower ID records as injection of hard coded overrides remains.

**Comment by RealHandcuffs** [comment]
I can confirm that modifying the header version solves the issue in FO4Edit.

I created a bug report after verifying that FO4Edit resolves things differently than the actual game. My testing showed clearly that the game can resolve the keyword, while FO4Edit did not resolve it, so I assumed a bug. If I understand your comments correctly then this is kind of an edge case (the game probably only resolve it because there is no record with ID 0x9); having divergent behavior in edge cases is probably ok, so this can be closed as NOT A BUG.

I will need to think about the best way of solving the issue. I am not sure if renumbering records will cause issues with existing savegames. It will also cause issues with mods that depend on my mod (there are a few), if these mods somehow reference one of the re-numerated records.

In any case, I thank you guys for the very quick answer 👍 .

**Comment by RealHandcuffs** [comment]
Sorry for the double post, I checked the situation again, it actually is like the following:

1. _RealHandcuffs.esp_ defines the keyword RH_Restraint (0x9). It also defines RH_Handcuffs (0x802). It has header version 1.0. 
2. _RealHandcuffs_DD_Compatibility.esp_ overrides RH_Handcuffs, carrying forward all keywords and adding new ones. One of these keywords that are carried forward is RH_Restraint. This plug-in has header version 0.95.

In other words I do not define a record less than 0x800 in a plug-in with version 0.95. I am only using it in one. Does this change the situation? Maybe this is the reason that the game can resolve the keyword...

**Comment by jbostrus** [comment]
> I will need to think about the best way of solving the issue. I am not sure if renumbering records will cause issues with existing savegames. It will also cause issues with mods that depend on my mod (there are a few), if these mods somehow reference one of the re-numerated records.

This is why I said:
> The least problem inducing is to change the HEDR version of plugins that involve that record to 1.0. The drawback here is that the users of the plugin need to be on a current game exe (1.10.162+) which then excludes VR.

This would be for any of *your* published plugins. If someone else references to your plugin as a master, then it is their responsibility to do likewise.

The only concern I would have at that point is for saves where the objects using that keyword may have been spawned prior to the HEDR change. I'm not sure how the engine will effectively react with that as it would appear to it as if a keyword was removed, and a new one added.

**Comment by jbostrus** [comment]
> In other words I do not define a record less than 0x800 in a plug-in with version 0.95. I am only using it in one. Does this change the situation? Maybe this is the reason that the game can resolve the keyword...

That actually should make it *not* resolve it. Because you define a keyword in one place. But your plugin is trying to access a non existent record. The only plausible reason I can think of as to why it may appear to be working, is if an object had previously been spawned with just the base plugin. Thus it has the keyword properly on it. Then a save made. You then add the patch plugin which has the broken link due to the HEDR 0.95. Then load the previous save. At this point the game loads the existing object which is still correct because it's using the keyword data from the state of the save. Spawning a new version of the object *should* result in it missing the keyword because it is now using the keyword data from the patch override.

Not all record values are applied to all objects during a save load.

**Comment by RealHandcuffs** [comment]
>The only plausible reason I can think of as to why it may appear to be working, is if an object had previously been spawned with just the base plugin.

Nope. Let me copy-paste [from my above link](https://www.loverslab.com/topic/118151-real-handcuffs/?do=findComment&comment=4001218):

Step 1: I added the following papyrus script to OnPlayerLoadGame:
```
RealHandcuffs:DebugWrapper.Trace(">>> RUNNING TEMPORARY CODE ON STARTUP <<<")
Armor handcuffs = Game.GetFormFromFile(0x000802, "RealHandcuffs.esp") as Armor
RealHandcuffs:DebugWrapper.Trace(">>> Handcuffs: " + LL_FourPlay.IntToHexString(handcuffs.GetFormID()) + " '" + handcuffs.GetName() + "'")
Game:PluginInfo[] plugins = Game.GetInstalledPlugins() 
Keyword[] kwds = handcuffs.GetKeywords()
RealHandcuffs:DebugWrapper.Trace(">>> Found " + kwds.Length + " keywords:")
Int index = 0
While (index < kwds.Length)
    Int formId = kwds[index].GetFormID()
    Int modId = formId / 0x01000000
    If (formId < 0)
        modId = 255 + modId ; assumes keyword is from not from light plugin
    EndIf
    RealHandcuffs:DebugWrapper.Trace(">>>   " + LL_FourPlay.IntToHexString(formId) + " (from " + plugins[modId].Name + ")")
    index += 1
EndWhile
RealHandcuffs:DebugWrapper.Trace(">>> FINISHED TEMPORARY CODE ON STARTUP <<<")
```

Step 2: With this code and both my mods (base plug-in and override plug-in) I started the game.

Step 3: I checked my papyrus log, it said:
```
>>> RUNNING TEMPORARY CODE ON STARTUP <<<
>>> Handcuffs: b3000802 'handcuffs'
>>> Found 8 keywords:
>>>   8f95b (from Fallout4.esm)
>>>   f4ae9 (from Fallout4.esm)
>>>   b300082a (from RealHandcuffs.esp)
>>>   b3000863 (from RealHandcuffs.esp)
>>>   1301196c (from Devious Devices.esm)
>>>   b3000965 (from RealHandcuffs.esp)
>>>   13004c5c (from Devious Devices.esm)
>>>   b3000009 (from RealHandcuffs.esp)
>>> FINISHED TEMPORARY CODE ON STARTUP <<<
```

- the line `1301196c (from Devious Devices.esm)` tells us that the game took the record from the patch (this keyword is added in the patch esp, it is missing in the base esp)
- the line `b3000009 (from RealHandcuffs.esp)` tells us that the game could in fact resolve the keyword

So again, the game can somehow resolve the keyword, while FO4Edit cannot. The question now basically is, did I just hit a lucky edge case and in general the game cannot resolve keywords like that, or is the logic of FO4Edit actually wrong and the game can resolve them even from a plugin with version 0.95, as long as the plugin that **defined** them had version 1.0?

Edit: There is no spawned instance of the armor. My papyrus script directly accesses the actual Armor record and uses papyrus functions to enumerate the keywords, so it shows the current state, not a previous state from the save game.

**Comment by ElminsterAU** [comment]
retest in 4.1.5
